### PR TITLE
Refactor navbar colour

### PR DESCRIPTION
**Pull request by obfusk**
Follow-up for #1559 (marked draft as it is based on that to avoid a merge conflict).

### CODE REVIEW

**Comment by TheLastProject** [comment]
I think you may be onto something here. I'm trying to take new screenshots and sometimes the bottom bar uses the wrong colour. Every now and then it goes right. There is indeed a race condition here...

**Comment by obfusk** [comment]
Rebased on #1559, no changes.

**Comment by obfusk** [comment]
> I think you may be onto something here.

Interesting. Have you tested if that still occurs with this change?

**Comment by obfusk** [comment]
It's ready IMO, but based on #1559 so marked as draft until that's merged. Though I guess you can merge this one and close the other instead if you prefer :)

**Comment by obfusk** [comment]
Oh yeah, I tested this in the emulator with Android 6 and 13. And on-device with Android 13. Seemed to work as expected.

**Comment by obfusk** [comment]
Rebased on `main`.

**Comment by obfusk** [comment]
This would make the code slightly more complicated, but OTOH I feel it's cleaner not to call `setNavigationBarColor()` twice. Not sure what's better.

```diff
diff --git a/app/src/main/java/protect/card_locker/CatimaAppCompatActivity.java b/app/src/main/java/protect/card_locker/CatimaAppCompatActivity.java
index 67b43a92..0ffb7458 100644
--- a/app/src/main/java/protect/card_locker/CatimaAppCompatActivity.java
+++ b/app/src/main/java/protect/card_locker/CatimaAppCompatActivity.java
@@ -14,6 +14,8 @@ import androidx.appcompat.app.AppCompatActivity;
 import androidx.core.view.WindowInsetsControllerCompat;
 
 public class CatimaAppCompatActivity extends AppCompatActivity {
+    protected boolean setNavBarColor = true;
+
     @Override
     protected void attachBaseContext(Context base) {
         // Apply chosen language
@@ -51,7 +53,9 @@ public class CatimaAppCompatActivity extends AppCompatActivity {
     @Override
     protected void onResume() {
         super.onResume();
-        Utils.setNavigationBarColor(this, null, Utils.resolveBackgroundColor(this), !Utils.isDarkModeEnabled(this));
+        if (setNavBarColor) {
+            Utils.setNavigationBarColor(this, null, Utils.resolveBackgroundColor(this), !Utils.isDarkModeEnabled(this));
+        }
     }
 
     protected void enableToolbarBackButton() {
diff --git a/app/src/main/java/protect/card_locker/LoyaltyCardViewActivity.java b/app/src/main/java/protect/card_locker/LoyaltyCardViewActivity.java
index 0944da40..a6c6ac78 100644
--- a/app/src/main/java/protect/card_locker/LoyaltyCardViewActivity.java
+++ b/app/src/main/java/protect/card_locker/LoyaltyCardViewActivity.java
@@ -553,6 +553,7 @@ public class LoyaltyCardViewActivity extends CatimaAppCompatActivity implements
 
     @Override
     protected void onResume() {
+        setNavBarColor = false;
         super.onResume();
 
         Log.i(TAG, "To view card: " + loyaltyCardId);
```

File: app/src/main/java/protect/card_locker/Utils.java [hunk 1]
```diff
@@ -669,21 +670,32 @@ public static void patchColors(AppCompatActivity activity) {
...
+        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.P) {
+            activity.findViewById(android.R.id.content).setBackgroundColor(resolveBackgroundColor(activity));
+        }
+    }
 
-        if (Build.VERSION.SDK_INT >= 27) {
-            Window window = activity.getWindow();
+    // Either pass an Activity on which to call getWindow() or an existing Window (may be null) returned by that function.
+    public static void setNavigationBarColor(@Nullable AppCompatActivity activity, @Nullable Window window, int color, boolean useLightBars) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
```
Comment (TheLastProject) [inline app/src/main/java/protect/card_locker/Utils.java:680]: We really need to start being consistent here and either use numbers or version codes, not both randomly.

I think most of the codebase uses ints but I have not strong opinion except for consistency.

File: app/src/main/java/protect/card_locker/Utils.java [hunk 2]
```diff
@@ -669,21 +670,32 @@ public static void patchColors(AppCompatActivity activity) {
     // rendering mess
     // use after views are inflated
     public static void postPatchColors(AppCompatActivity activity) {
-        TypedValue typedValue = new TypedValue();
-        activity.getTheme().resolveAttribute(android.R.attr.colorBackground, typedValue, true);
-        activity.findViewById(android.R.id.content).setBackgroundColor(typedValue.data);
+        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.P) {
+            activity.findViewById(android.R.id.content).setBackgroundColor(resolveBackgroundColor(activity));
+        }
```
Comment (TheLastProject) [inline app/src/main/java/protect/card_locker/Utils.java:675]: Please use ints as that's what most of Catima users. If you want to change the standard, let's do that in another MR (and let's start documenting these things somewhere).

And you say "I tested this in the emulator with Android 6 and 13.", I'm most curious actually about the versions that no longer run this code. So, 10 and 11 and 12 and 14 too.

So this needs a lot of testing on multiple different themes (that I don't have time to do right now).

**Review by TheLastProject — CHANGES_REQUESTED** [review]

File: app/src/main/java/protect/card_locker/Utils.java (hunk 2 above)
Comment (obfusk) [reply app/src/main/java/protect/card_locker/Utils.java:675]: > Please use ints as that's what most of Catima users.

Does it? See above.

I'll look into testing more. No rush :)

File: app/src/main/java/protect/card_locker/Utils.java (hunk 1 above)
Comment (obfusk) [reply app/src/main/java/protect/card_locker/Utils.java:680]: I can never remember what Android version or code the numbers map to. At least with `Build.VERSION_CODES.O_MR1` I get a pop-up in Android Studio telling me that's "Android 8.1" and its value is `27`, giving me the version code, version, and level all at once. That's much nicer than hardcoding `27` as I would need to manually look up which android version & code that is. Of course we could add comments for that instead if we really want to hardcode versions, but if we change those we might forget to keep the comment in sync.

I do agree on consistency. And when I looked at the existing codebase I found more version code constants than hardcoded numbers. Happy to follow up with a PR to use version codes in the two remaining places as well.

```
$ ack Build.VERSION.SDK_INT
app/src/main/java/protect/card_locker/CardsOnPowerScreenService.java
143:        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.R) {

app/src/main/java/protect/card_locker/CatimaAppCompatActivity.java
37:            if (Build.VERSION.SDK_INT >= 23) {

app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java
206:    // store system locale for Build.VERSION.SDK_INT < Build.VERSION_CODES.N
468:                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {

app/src/main/java/protect/card_locker/LoyaltyCardViewActivity.java
582:                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
637:        if (window != null && Build.VERSION.SDK_INT >= 27) {
1064:            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
1098:            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {

app/src/main/java/protect/card_locker/PermissionUtils.java
23:        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
38:        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {

app/src/main/java/protect/card_locker/UCropWrapper.java
33:        if (Build.VERSION.SDK_INT >= 23) {

app/src/main/java/protect/card_locker/Utils.java
205:        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
487:        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
505:        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
676:        if (Build.VERSION.SDK_INT >= 27) {
765:            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {

app/src/main/java/protect/card_locker/contentprovider/CardsContentProvider.java
79:        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {

app/src/main/java/protect/card_locker/preferences/SettingsActivity.java
155:            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
181:                if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
195:            contentProviderReadPreference.setVisible(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
```

Comment (TheLastProject) [reply app/src/main/java/protect/card_locker/Utils.java:680]: Okay those are very good argument in favour of always using the codes :)

File: app/src/main/java/protect/card_locker/Utils.java (hunk 2 above)
Comment (TheLastProject) [reply app/src/main/java/protect/card_locker/Utils.java:675]: Well, nvm, just more testing then :)

Comment (TheLastProject) [reply app/src/main/java/protect/card_locker/Utils.java:675]: Turns out this code is responsible for correctly applying the "Pure black background for dark theme" setting when dark theme is used in the MainActivity. Adding the version check breaks the OLED black on new Android versions (noticed on Android 14)

Strangely enough, this code isn't necessary in the LoyaltyCardEditActivity, what a mess...

Comment (obfusk) [reply app/src/main/java/protect/card_locker/Utils.java:675]: Will do more testing & debugging. Hopefully this week.

Comment (TheLastProject) [reply app/src/main/java/protect/card_locker/Utils.java:675]: Maybe it'd be best to just leave this code on all API versions with a new comment like:

```
// Force correct color
// Fixes OLED dark mode in MainActivity
```

So we at least understand better what it does. And then later debug further

**Pull request merged**